crates/amaru-ledger/src/rules/transaction/scripts.rs

```diff
@@ -1,59 +1,115 @@
 use std::collections::BTreeSet;
 
 use amaru_kernel::{
-    get_provided_scripts, HasScriptHash, MintedWitnessSet, ScriptHash, TransactionInput,
+    get_provided_scripts, DatumOption, HasAddress, HasDatum, HasScriptHash, Hash, MintedWitnessSet,
+    OriginalHash, PseudoScript, ScriptHash, ScriptRefWithHash, TransactionInput,
 };
 use thiserror::Error;
 
 use crate::context::{UtxoSlice, WitnessSlice};
 
 #[derive(Debug, Error)]
 pub enum InvalidScripts {
-    #[error("missing required scripts: missing {0:?}")]
+    #[error("missing required scripts: missing [{}]",
+        .0.iter().map(|hash| hash.to_string()).collect::<Vec<_>>().join(", "),
+    )]
     MissingRequiredScripts(Vec<ScriptHash>),
-    #[error("extraneous script witnesses: extra {0:?}")]
+    #[error("extraneous script witnesses: extra [{}]",
+        .0.iter().map(|hash| hash.to_string()).collect::<Vec<_>>().join(", "),
+    )]
     ExtraneousScriptWitnesses(Vec<ScriptHash>),
+    #[error("unspendable inputs; no datums: [{}]",
+        .0
+        .iter()
+        .map(|input|
+            format!("{}#{}", input.transaction_id, input.index)
+        )
+        .collect::<Vec<_>>()
+        .join(", ")
+    )]
+    UnspendableInputsNoDatums(Vec<TransactionInput>),
+    #[error(
+        "missing required datums: missing [{}] provided [{}]",
+        missing.iter().map(|hash| hash.to_string()).collect::<Vec<_>>().join(", "),
+        provided.iter().map(|hash| hash.to_string()).collect::<Vec<_>>().join(", "),
+    )]
+    MissingRequiredDatums {
+        missing: Vec<Hash<32>>,
+        provided: BTreeSet<Hash<32>>,
+    },
+    #[error(
+        "extraneous supplemental datums: allowed: [{}] provided [{}]",
+        allowed.iter().map(|hash| hash.to_string()).collect::<Vec<_>>().join(", "),
+        provided.iter().map(|hash| hash.to_string()).collect::<Vec<_>>().join(", "),
+    )]
+    ExtraneousSupplementalDatums {
+        allowed: BTreeSet<Hash<32>>,
+        provided: BTreeSet<Hash<32>>,
+    },
 }
 
+// TODO: this can be made MUCH more efficient. Remove clones, don't iterate the same list several times, etc... Lots of low hanging fruit.
 pub fn execute<C>(
     context: &mut C,
     reference_inputs: Option<&Vec<TransactionInput>>,
-    inputs: &Vec<TransactionInput>,
+    inputs: &[TransactionInput],
```

Comment:
I don't think these guys should be here as they've already been processed during the `inputs` rules. Rather; while processing inputs (and outputs), we shall register scripts and datums that are found. In the same way we do it for certificates and credentials.

So this particular rule could have immediate access to the scripts present in the transaction body. It may only need the script witness as an argument (and not the whole witness since we check vk witnesses elsewhere).

Reply:
I agree -- but I think that this will be addressed in future PRs. I've coupled 3 different validation rules together for the `scripts.rs` file, and I'm working on finishing up the redeemers now. My plan was to go back and improve all of this once I have them all written up with tests, so it's easier to tell what I'm breaking.